src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.cs

```
@@ -350,7 +350,14 @@ internal virtual bool IsPatternSupported(UiaCore.UIA patternId)
             return false;
         }
 
-        internal virtual int[]? RuntimeId => null;
+        internal virtual int[] RuntimeId
```

I think, @Tanya-Solyanik suggested throwing a IOE or NIE here.

---

Just throwing an exception is not enough to be really noticeable. Instead I'm using `Debug.Fail` to break VS in the debugging mode, also returning an empty array in case of non-debug builds. It is also possible to throw an exception instead of returning an empty array, but this slows down accessibility tools a bit in case of thrown exceptions.

---

The exception is here to alert developers that they haven't correctly implement accessibility support. So in the general case no one should see the exception.
Is this not the case?

---

@RussKie the problem is that an exception doesn't break the debugging in this case. It is hard to notice it as it appears only in the output and even without the stacktrace.

---

Ping me, let's chat.

---

Discussed this with @RussKie: instead of `Debug.Fail` we will throw an exception (e.g. `NotImplementedException`) and add a test to check the `RuntimeId` property of all the descendants of the `AccessibilityObject` to not throw the exception from the base class.

---

It would be nice to add description on why we went with exception here given that it slows inspect tools. @WPMGPRoSToTeMa.

---

@RussKie I've reverted `Debug.Fail`. This won't affect the end users as we're not providing debug dlls to them. Though I've added the test to check that all classes override `RuntimeId` it is still useful to have an assertion because a developer may override it as `base.RuntimeId` accidentally.

---

I think that not implemented exception is more descriptive - it suggests to implement this method in the derived class.

---

@Tanya-Solyanik looks like it is really better to throw `NotImplementedException` than `InvalidOperationException` since the latter should be thrown in cases when the object state doesn't allow using this property/method (in our case the base property cannot be used all the time):
`This exception is thrown in scenarios in which it is generally sometimes possible for the object to perform the requested operation, and the object state determines whether the operation can be performed.`
https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=net-5.0#related-exception-types

There is also `NotSupportedException`:
`You should also throw a NotSupportedException exception if you've implemented an abstract base class and add a new member to it that must be overridden by derived classes. In that case, making the member abstract causes existing subclasses to fail to load.`
https://docs.microsoft.com/en-us/dotnet/api/system.notimplementedexception?view=net-5.0#notimplementedexception-and-other-exception-types

A note about `NotImplementedException`:
`You might choose to throw a NotImplementedException exception in properties or methods in your own types when the that member is still in development and will only later be implemented in production code. In other words, a NotImplementedException exception should be synonymous with "still in development."`
https://docs.microsoft.com/en-us/dotnet/api/system.notimplementedexception?view=net-5.0#throwing-the-exception

I think `NotImplementedException` may confuse a person who will read the code, because it looks like it will be implemented later and in this case `NotSupportedException` looks more preferably. But if there is an assertion (`Debug.Fail`) or a comment it won't confuse a reader and therefore `NotImplementedException` can be used without any confusion. So I think both are applicable in this case, but I will use `NotImplementedException` as you suggested.

---

>may confuse a person who will read the code, because it looks like it will be implemented later (and it is possible to implement it) 

Sure, NotImplemented looks like code is incomplete, we can't make it abstract at this point, NotSupportedException is the next best thing.

Please localize your message - 
>const string message = $"{nameof(AccessibleObject)} doesn't support {nameof(RuntimeId)} property. It must be overriden in the derived class.";